engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java

```diff
@@ -1738,38 +1738,58 @@ public Class<?> visit(MethodCallExpr n, VisitArgs printer) {
         return calculateMethodReturnTypeUsingGenerics(scope, method, expressionTypes, parameterizedTypes);
     }
 
-    private void checkPyNumbaVectorizedFunc(MethodCallExpr n, Expression[] expressions, Class<?>[] expressionTypes) {
-        // numba vectorized functions return arrays of primitive types. This will break the generated expression
+    private void prepareVectorization(MethodCallExpr n, Expression[] expressions, Class<?>[] expressionTypes) {
+        // Python vectorized functions(numba, DH) return arrays of primitive types. This will break the generated
+        // expression
         // evaluation code that expects singular values. This check makes sure that numba vectorized functions must be
         // used alone (or with cast only) as the entire expression.
-        n.getParentNode().ifPresent(parent -> {
-            if (parent.getClass() != CastExpr.class || parent.getParentNode().isPresent()) {
-                throw new RuntimeException("Numba vectorized function can't be used in an expression.");
-            }
-        });
 
         final QueryScope queryScope = ExecutionContext.getContext().getQueryScope();
         for (QueryScopeParam<?> param : queryScope.getParams(queryScope.getParamNames())) {
             if (param.getName().equals(n.getNameAsString())) {
-                NumbaCallableWrapper numbaCallableWrapper = (NumbaCallableWrapper) param.getValue();
-                List<Class<?>> params = numbaCallableWrapper.getParamTypes();
-                if (params.size() != expressions.length) {
-                    throw new RuntimeException("Numba vectorized function argument count mismatch: " + params.size()
-                            + " vs." + expressions.length);
-                }
-                for (int i = 0; i < expressions.length; i++) {
-                    if (!(expressions[i] instanceof NameExpr)) {
-                        throw new RuntimeException("Numba vectorized function arguments can only be columns.");
-                    }
-                    if (!isSafelyCoerceable(expressionTypes[i], params.get(i))) {
-                        throw new RuntimeException("Numba vectorized function argument type mismatch: "
-                                + expressionTypes[i].getSimpleName() + " -> " + params.get(i).getSimpleName());
-                    }
+                CallableWrapper callableWrapper = (CallableWrapper) param.getValue();
+                try {
+                    checkVectorizability(n, expressions, expressionTypes, callableWrapper);
+                } catch (RuntimeException ex) {
+                    if (callableWrapper.isVectorized())
+                        throw ex;
                 }
```

Comment:
Added comments. We only parse the formula once, so for functions vectorized by hand (decorated with @numba.vectorized or @dh_vectorized), this would be a compilation error. However for a plain Python function we simply deem it as un-vectorizable and will not auto vectorize it later in the evaluation process.